{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load allthebacteria metadata (what seems to be the most interesting file for now)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_339313/3982220849.py:10: DtypeWarning: Columns (5,6,8,9,10,11,12,13,14,15,20,21,22,25,26,27,28,30,31,36,39,41,42,46,47,48,51,52,55,56,58,59,60,62,63,64,66,68,69,70,72,73,77,79,80,83,89,90,91,92,99,100,102,103,112,115,116) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(content, sep=\"\\t\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>assembly_software</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>serotype</th>\n",
       "      <th>environment_feature</th>\n",
       "      <th>last_updated</th>\n",
       "      <th>submitted_galaxy</th>\n",
       "      <th>germline</th>\n",
       "      <th>...</th>\n",
       "      <th>taxonomic_classification</th>\n",
       "      <th>protocol_label</th>\n",
       "      <th>elevation</th>\n",
       "      <th>salinity</th>\n",
       "      <th>sequencing_method</th>\n",
       "      <th>first_public</th>\n",
       "      <th>study_alias</th>\n",
       "      <th>ph</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>isolation_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMD00008604</td>\n",
       "      <td>DRR008449</td>\n",
       "      <td>DRA000969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2015-06-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014-04-28</td>\n",
       "      <td>DRP001007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMD00002817</td>\n",
       "      <td>DRR016177</td>\n",
       "      <td>DRA001721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-10-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-10-02</td>\n",
       "      <td>PRJDB1847</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMD00019027</td>\n",
       "      <td>DRR021788</td>\n",
       "      <td>DRA002419</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-10-21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-10-21</td>\n",
       "      <td>PRJDB2736</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMD00019237</td>\n",
       "      <td>DRR022206</td>\n",
       "      <td>DRA002431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-12-28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-12-28</td>\n",
       "      <td>DRP008013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMD00019241</td>\n",
       "      <td>DRR022210</td>\n",
       "      <td>DRA002431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-12-28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-12-28</td>\n",
       "      <td>DRP008013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1943488</th>\n",
       "      <td>SAMN10405967</td>\n",
       "      <td>SRR8179222</td>\n",
       "      <td>SRA808114</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>Heidelberg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-11-10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-11-10</td>\n",
       "      <td>PRJNA242847</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>comminuted beef</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1943489</th>\n",
       "      <td>SAMN10405968</td>\n",
       "      <td>SRR8179219</td>\n",
       "      <td>SRA808111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>Anatum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-11-10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-11-10</td>\n",
       "      <td>PRJNA242847</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>comminuted beef</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1943490</th>\n",
       "      <td>SAMN10407504</td>\n",
       "      <td>SRR8180444</td>\n",
       "      <td>SRA808182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>not typed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-11-10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-11-10</td>\n",
       "      <td>PRJNA212117</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1943491</th>\n",
       "      <td>SAMN10407505</td>\n",
       "      <td>SRR8180405</td>\n",
       "      <td>SRA808181</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>not typed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-11-10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-11-10</td>\n",
       "      <td>PRJNA212117</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>placenta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1943492</th>\n",
       "      <td>SAMN10407507</td>\n",
       "      <td>SRR8180452</td>\n",
       "      <td>SRA808179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>not typed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-11-10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-11-10</td>\n",
       "      <td>PRJNA212117</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1943493 rows × 118 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample_accession run_accession submission_accession  \\\n",
       "0           SAMD00008604     DRR008449            DRA000969   \n",
       "1           SAMD00002817     DRR016177            DRA001721   \n",
       "2           SAMD00019027     DRR021788            DRA002419   \n",
       "3           SAMD00019237     DRR022206            DRA002431   \n",
       "4           SAMD00019241     DRR022210            DRA002431   \n",
       "...                  ...           ...                  ...   \n",
       "1943488     SAMN10405967    SRR8179222            SRA808114   \n",
       "1943489     SAMN10405968    SRR8179219            SRA808111   \n",
       "1943490     SAMN10407504    SRR8180444            SRA808182   \n",
       "1943491     SAMN10407505    SRR8180405            SRA808181   \n",
       "1943492     SAMN10407507    SRR8180452            SRA808179   \n",
       "\n",
       "         assembly_software   library_selection    serotype  \\\n",
       "0                      NaN              RANDOM         NaN   \n",
       "1                      NaN  size fractionation         NaN   \n",
       "2                      NaN              RANDOM         NaN   \n",
       "3                      NaN              RANDOM         NaN   \n",
       "4                      NaN              RANDOM         NaN   \n",
       "...                    ...                 ...         ...   \n",
       "1943488                NaN              RANDOM  Heidelberg   \n",
       "1943489                NaN              RANDOM      Anatum   \n",
       "1943490                NaN              RANDOM   not typed   \n",
       "1943491                NaN              RANDOM   not typed   \n",
       "1943492                NaN              RANDOM   not typed   \n",
       "\n",
       "        environment_feature last_updated submitted_galaxy germline  ...  \\\n",
       "0                       NaN   2015-06-19              NaN      NaN  ...   \n",
       "1                       NaN   2018-10-02              NaN      NaN  ...   \n",
       "2                       NaN   2017-10-21              NaN      NaN  ...   \n",
       "3                       NaN   2021-12-28              NaN      NaN  ...   \n",
       "4                       NaN   2021-12-28              NaN      NaN  ...   \n",
       "...                     ...          ...              ...      ...  ...   \n",
       "1943488                 NaN   2018-11-10              NaN    False  ...   \n",
       "1943489                 NaN   2018-11-10              NaN    False  ...   \n",
       "1943490                 NaN   2018-11-10              NaN    False  ...   \n",
       "1943491                 NaN   2018-11-10              NaN    False  ...   \n",
       "1943492                 NaN   2018-11-10              NaN    False  ...   \n",
       "\n",
       "        taxonomic_classification protocol_label elevation salinity  \\\n",
       "0                            NaN            NaN       NaN      NaN   \n",
       "1                            NaN            NaN       NaN      NaN   \n",
       "2                            NaN            NaN       NaN      NaN   \n",
       "3                            NaN            NaN       NaN      NaN   \n",
       "4                            NaN            NaN       NaN      NaN   \n",
       "...                          ...            ...       ...      ...   \n",
       "1943488                      NaN            NaN       NaN      NaN   \n",
       "1943489                      NaN            NaN       NaN      NaN   \n",
       "1943490                      NaN            NaN       NaN      NaN   \n",
       "1943491                      NaN            NaN       NaN      NaN   \n",
       "1943492                      NaN            NaN       NaN      NaN   \n",
       "\n",
       "        sequencing_method first_public  study_alias   ph  tissue_type  \\\n",
       "0                     NaN   2014-04-28    DRP001007  NaN          NaN   \n",
       "1                     NaN   2018-10-02    PRJDB1847  NaN          NaN   \n",
       "2                     NaN   2017-10-21    PRJDB2736  NaN          NaN   \n",
       "3                     NaN   2021-12-28    DRP008013  NaN          NaN   \n",
       "4                     NaN   2021-12-28    DRP008013  NaN          NaN   \n",
       "...                   ...          ...          ...  ...          ...   \n",
       "1943488               NaN   2018-11-10  PRJNA242847  NaN          NaN   \n",
       "1943489               NaN   2018-11-10  PRJNA242847  NaN          NaN   \n",
       "1943490               NaN   2018-11-10  PRJNA212117  NaN          NaN   \n",
       "1943491               NaN   2018-11-10  PRJNA212117  NaN          NaN   \n",
       "1943492               NaN   2018-11-10  PRJNA212117  NaN          NaN   \n",
       "\n",
       "         isolation_source  \n",
       "0                     NaN  \n",
       "1                     NaN  \n",
       "2                     NaN  \n",
       "3                     NaN  \n",
       "4                     NaN  \n",
       "...                   ...  \n",
       "1943488   comminuted beef  \n",
       "1943489   comminuted beef  \n",
       "1943490             blood  \n",
       "1943491          placenta  \n",
       "1943492               NaN  \n",
       "\n",
       "[1943493 rows x 118 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "MD_PATH = Path(\"/home/cyrille/dev/microtaxo/data/metadata\")\n",
    "\n",
    "\n",
    "# content = MD_PATH / \"assembly-stats.decontam.tsv\"\n",
    "content = MD_PATH / \"ena_metadata.tsv\"\n",
    "df = pd.read_csv(content, sep=\"\\t\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check columns names/values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sample_accession': 'SAMD00008604',\n",
       " 'run_accession': 'DRR008449',\n",
       " 'submission_accession': 'DRA000969',\n",
       " 'assembly_software': nan,\n",
       " 'library_selection': 'RANDOM',\n",
       " 'serotype': nan,\n",
       " 'environment_feature': nan,\n",
       " 'last_updated': '2015-06-19',\n",
       " 'submitted_galaxy': nan,\n",
       " 'germline': nan,\n",
       " 'culture_collection': nan,\n",
       " 'submission_tool': nan,\n",
       " 'sra_bytes': 130004514.0,\n",
       " 'collected_by': nan,\n",
       " 'submitted_ftp': nan,\n",
       " 'isolate': nan,\n",
       " 'fastq_bytes': '132346864;79650546',\n",
       " 'instrument_platform': 'ILLUMINA',\n",
       " 'variety': nan,\n",
       " 'temperature': nan,\n",
       " 'sra_aspera': 'fasp.sra.ebi.ac.uk:/vol1/drr/DRR008/DRR008449',\n",
       " 'ecotype': nan,\n",
       " 'submitted_aspera': nan,\n",
       " 'sampling_campaign': nan,\n",
       " 'tissue_lib': nan,\n",
       " 'environmental_sample': nan,\n",
       " 'sex': nan,\n",
       " 'submitted_md5': nan,\n",
       " 'checklist': nan,\n",
       " 'fastq_galaxy': 'ftp.sra.ebi.ac.uk/vol1/fastq/DRR008/DRR008449/DRR008449_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/DRR008/DRR008449/DRR008449_2.fastq.gz',\n",
       " 'specimen_voucher': nan,\n",
       " 'submitted_bytes': nan,\n",
       " 'taxonomic_identity_marker': nan,\n",
       " 'country': nan,\n",
       " 'sample_description': '168-B',\n",
       " 'experiment_title': 'Illumina Genome Analyzer IIx sequencing; Whole genome re-sequencing of Bacillus subtilis 168 from the laboratory B;Whole genome re-sequencing of Bacillus subtilis 168 from the laboratory B',\n",
       " 'sra_galaxy': 'ftp.sra.ebi.ac.uk/vol1/drr/DRR008/DRR008449',\n",
       " 'fastq_md5': 'f2ade9a77872c9506f74bd98c39ef567;728c363d6f7b5833e6fcafd74b6e7039',\n",
       " 'secondary_study_accession': 'DRP001007',\n",
       " 'read_count': 2062118,\n",
       " 'study_title': 'Resequence Analysis of B. subtilis laboratory strains',\n",
       " 'bio_material': nan,\n",
       " 'host_body_site': nan,\n",
       " 'assembly_quality': nan,\n",
       " 'sample_capture_status': nan,\n",
       " 'sample_title': '168-B',\n",
       " 'host_genotype': nan,\n",
       " 'host_phenotype': nan,\n",
       " 'cultivar': nan,\n",
       " 'instrument_model': 'Illumina Genome Analyzer IIx',\n",
       " 'target_gene': nan,\n",
       " 'nominal_sdev': nan,\n",
       " 'environment_material': nan,\n",
       " 'host_tax_id': nan,\n",
       " 'sample_material': nan,\n",
       " 'sra_md5': '5c46681343408da7031a5dca07a21573',\n",
       " 'cell_type': nan,\n",
       " 'fastq_ftp': 'ftp.sra.ebi.ac.uk/vol1/fastq/DRR008/DRR008449/DRR008449_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/DRR008/DRR008449/DRR008449_2.fastq.gz',\n",
       " 'broker_name': nan,\n",
       " 'sub_strain': nan,\n",
       " 'base_count': 309317700,\n",
       " 'library_strategy': 'WGS',\n",
       " 'serovar': nan,\n",
       " 'investigation_type': nan,\n",
       " 'location': nan,\n",
       " 'library_source': 'GENOMIC',\n",
       " 'sra_ftp': 'ftp.sra.ebi.ac.uk/vol1/drr/DRR008/DRR008449',\n",
       " 'library_layout': 'PAIRED',\n",
       " 'experimental_factor': nan,\n",
       " 'environment_biome': nan,\n",
       " 'dev_stage': nan,\n",
       " 'binning_software': nan,\n",
       " 'sample_collection': nan,\n",
       " 'nominal_length': nan,\n",
       " 'completeness_score': nan,\n",
       " 'lon': nan,\n",
       " 'fastq_aspera': 'fasp.sra.ebi.ac.uk:/vol1/fastq/DRR008/DRR008449/DRR008449_1.fastq.gz;fasp.sra.ebi.ac.uk:/vol1/fastq/DRR008/DRR008449/DRR008449_2.fastq.gz',\n",
       " 'host_sex': nan,\n",
       " 'sample_alias': 'SAMD00008604',\n",
       " 'mating_type': nan,\n",
       " 'sub_species': 'subtilis',\n",
       " 'contamination_score': nan,\n",
       " 'run_alias': 'DRR008449',\n",
       " 'depth': nan,\n",
       " 'host_growth_conditions': nan,\n",
       " 'collection_date': nan,\n",
       " 'experiment_alias': 'DRX007637',\n",
       " 'host_gravidity': nan,\n",
       " 'center_name': 'TUAGRI',\n",
       " 'identified_by': nan,\n",
       " 'cell_line': nan,\n",
       " 'sampling_site': nan,\n",
       " 'host': nan,\n",
       " 'library_name': '168-B',\n",
       " 'first_created': '2014-04-28',\n",
       " 'lat': nan,\n",
       " 'strain': '168',\n",
       " 'experiment_accession': 'DRX007637',\n",
       " 'scientific_name': 'Bacillus subtilis subsp. subtilis str. 168',\n",
       " 'host_status': nan,\n",
       " 'tax_id': 224308.0,\n",
       " 'study_accession': 'PRJDB1076',\n",
       " 'submitted_format': nan,\n",
       " 'submitted_host_sex': nan,\n",
       " 'altitude': nan,\n",
       " 'accession': 'DRR008449',\n",
       " 'secondary_sample_accession': 'DRS007578',\n",
       " 'sampling_platform': nan,\n",
       " 'taxonomic_classification': nan,\n",
       " 'protocol_label': nan,\n",
       " 'elevation': nan,\n",
       " 'salinity': nan,\n",
       " 'sequencing_method': nan,\n",
       " 'first_public': '2014-04-28',\n",
       " 'study_alias': 'DRP001007',\n",
       " 'ph': nan,\n",
       " 'tissue_type': nan,\n",
       " 'isolation_source': nan}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[0].to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## basic exploration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "118"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([nan])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# nothing there\n",
    "df[\"taxonomic_classification\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([nan])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# nothing there\n",
    "df[\"taxonomic_identity_marker\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([224308.0, 1280.0, 562.0, ..., 2486975, 2486976, 2486977],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# looks interesting\n",
    "df[\"tax_id\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Bacillus subtilis subsp. subtilis str. 168',\n",
       "       'Staphylococcus aureus', 'Escherichia coli', ...,\n",
       "       'Campylobacter sp. OH-18-19611-1',\n",
       "       'Campylobacter sp. OH-18-19102-CU',\n",
       "       'Campylobacter sp. OH-18-15554-1'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"scientific_name\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## populate API cache"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/cyrille/.cache/pypoetry/virtualenvs/microtaxo-SPpTyrCZ-py3.10/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "100%|██████████| 38659/38659 [00:01<00:00, 21667.00it/s]\n"
     ]
    }
   ],
   "source": [
    "from urllib.error import HTTPError\n",
    "from Bio import Entrez\n",
    "from diskcache import Cache\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "Entrez.email = \"cyrille.leroux@irisa.fr\"\n",
    "CACHE_DIR = \"/data/microtaxo/apicache\"\n",
    "cache = Cache(CACHE_DIR)\n",
    "\n",
    "# ~4-5h without API key\n",
    "errors = list()\n",
    "unique_tax_ids = list()\n",
    "\n",
    "def get_api_data(tax_id):\n",
    "    # error: not a number / None\n",
    "    if pd.isna(tax_id):\n",
    "        errors.append(tax_id)\n",
    "        return None      \n",
    "    \n",
    "    # float/string -> int\n",
    "    try:        \n",
    "        tax_id = int(tax_id)\n",
    "    except ValueError:\n",
    "        errors.append(tax_id)\n",
    "        return None\n",
    "    \n",
    "    # hit cache\n",
    "    if tax_id in cache:\n",
    "        unique_tax_ids.append(tax_id)\n",
    "        return cache[tax_id]\n",
    "    \n",
    "     # API call + cache\n",
    "    try:\n",
    "        handle = Entrez.efetch(db=\"taxonomy\", id=str(tax_id), retmode=\"xml\")\n",
    "        records = Entrez.read(handle)\n",
    "        cache[tax_id] = records\n",
    "        unique_tax_ids.append(tax_id)\n",
    "        return records\n",
    "    except HTTPError:\n",
    "        errors.append(tax_id)\n",
    "\n",
    "    return None\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "for tax_id in tqdm(df[\"tax_id\"].unique()):\n",
    "    get_api_data(tax_id)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 38511/38511 [00:01<00:00, 20910.80it/s]\n",
      "100%|██████████| 6566/6566 [00:00<00:00, 19402.52it/s]\n"
     ]
    }
   ],
   "source": [
    "# extra tax_id\n",
    "extra_tax_ids = set()\n",
    "for tax_id in tqdm(unique_tax_ids):\n",
    "    if tax_id in cache:\n",
    "        if tdata := cache[tax_id]:\n",
    "            extra_tax_ids.update([int(lin[\"TaxId\"]) for lin in tdata[0].get(\"LineageEx\", {})])\n",
    "for tax_id in tqdm(extra_tax_ids):\n",
    "    get_api_data(tax_id)\n",
    "extra_tax_ids = list(extra_tax_ids)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### let's see what wrong with rejected tax_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(148,\n",
       " 45077,\n",
       " 6566,\n",
       " [nan,\n",
       "  '54736,33035',\n",
       "  '33035,83558',\n",
       "  '1313,561276',\n",
       "  '561276,1313',\n",
       "  '28449,486',\n",
       "  '34026,486',\n",
       "  '34026,488',\n",
       "  '488,1906581',\n",
       "  '33053,488',\n",
       "  '267212,488',\n",
       "  '486,495',\n",
       "  '488,28449',\n",
       "  '488,28091',\n",
       "  '488,267212',\n",
       "  '33053,486',\n",
       "  '483,486',\n",
       "  '28449,490',\n",
       "  '28449,488',\n",
       "  '482,486',\n",
       "  '194197,486',\n",
       "  '489,28449',\n",
       "  '28449,34026',\n",
       "  '28449,33053',\n",
       "  '488,486',\n",
       "  '493,486',\n",
       "  '488,483',\n",
       "  '489,483',\n",
       "  '28449,483',\n",
       "  '90368,483',\n",
       "  '483,489',\n",
       "  '727,726',\n",
       "  '729,727',\n",
       "  '727,729',\n",
       "  '729,726',\n",
       "  '726,729',\n",
       "  '119912,57046',\n",
       "  '119912,590',\n",
       "  '41529,119912',\n",
       "  '41529,590',\n",
       "  '57046,119912',\n",
       "  '590,41529',\n",
       "  '273136,813',\n",
       "  '813,273136',\n",
       "  '149539,98360',\n",
       "  '98360,149539',\n",
       "  '1781,1806',\n",
       "  '1806,1781',\n",
       "  '1773,33894',\n",
       "  '33894,1773',\n",
       "  '1765,1773',\n",
       "  '818,823',\n",
       "  '823,821',\n",
       "  '363265,823',\n",
       "  '29485,33060',\n",
       "  '33060,29485',\n",
       "  '631,633',\n",
       "  '633,631',\n",
       "  '633,29484',\n",
       "  '633,33060',\n",
       "  '29486,633',\n",
       "  '630,633',\n",
       "  '633,28152',\n",
       "  '631,29484',\n",
       "  '29485,29484',\n",
       "  '631,29485',\n",
       "  '29484,634',\n",
       "  '29484,33060',\n",
       "  '1288385,29484',\n",
       "  '631,28152',\n",
       "  '33060,631',\n",
       "  '29485,631',\n",
       "  '263819,633',\n",
       "  '28152,29484',\n",
       "  '33060,29484',\n",
       "  '29484,29485',\n",
       "  '634,29484',\n",
       "  '633,367190',\n",
       "  '367190,633',\n",
       "  '28152,633',\n",
       "  '29484,419257',\n",
       "  '263819,29484',\n",
       "  '28152,631',\n",
       "  '29484,631',\n",
       "  '631,33060',\n",
       "  '633,419257',\n",
       "  '630,29484',\n",
       "  '1314,119602',\n",
       "  '119602,1314',\n",
       "  '244366,573',\n",
       "  '573,244366',\n",
       "  '1463165,573',\n",
       "  '573,1463165',\n",
       "  '90370,54388',\n",
       "  '54388,90370',\n",
       "  '54388,57046',\n",
       "  '57046,54388',\n",
       "  '813,470',\n",
       "  '813,83555',\n",
       "  '83555,813',\n",
       "  '149387,149386',\n",
       "  '82689,28147',\n",
       "  '260677,28147',\n",
       "  '913076,28147',\n",
       "  '28147,70803',\n",
       "  '28147,149386',\n",
       "  '363569,28147',\n",
       "  '82689,115981',\n",
       "  '165302,115981',\n",
       "  '486988,115981',\n",
       "  '115981,28147',\n",
       "  '149387,115981',\n",
       "  '1160751,115981',\n",
       "  '363569,115981',\n",
       "  '1160747,149386',\n",
       "  '486993,149386',\n",
       "  '149386,363569',\n",
       "  '149386,913076',\n",
       "  '486988,363569',\n",
       "  '363569,149386',\n",
       "  '1151002,363569',\n",
       "  '363569,165302',\n",
       "  '363569,29472',\n",
       "  '115981,363569',\n",
       "  '82689,363569',\n",
       "  '29472,28147',\n",
       "  '149387,28147',\n",
       "  '1160739,28147',\n",
       "  '224729,57045',\n",
       "  '57045,224729',\n",
       "  '54736,28901',\n",
       "  '595,149385',\n",
       "  '149385,595',\n",
       "  '195,197',\n",
       "  '197,195',\n",
       "  '1813019,197',\n",
       "  '706436,762963',\n",
       "  '706434,749489',\n",
       "  '749496,749492,706433',\n",
       "  '749491,748224,749497',\n",
       "  '749529,749535',\n",
       "  '762966,762968,626939',\n",
       "  '749537,749527',\n",
       "  '765065,765070,765105,765096',\n",
       "  '765103,765084,765075,765115,765111,765085',\n",
       "  '765066,765117,765068,765107,765112',\n",
       "  '765072,765097',\n",
       "  '765069,765118'])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# it looks like multiple values in single field: we may or may not do something about it later\n",
    "len(errors), len(unique_tax_ids), len(extra_tax_ids), errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CreateDate': '2002/05/08 12:00:00',\n",
      " 'Division': 'Bacteria',\n",
      " 'GeneticCode': {'GCId': '11', 'GCName': 'Bacterial, Archaeal and Plant Plastid'},\n",
      " 'Lineage': 'cellular organisms; Bacteria; Bacillati; Bacillota; Bacilli; Bacillales; Bacillaceae; Bacillus; Bacillus subtilis group; Bacillus subtilis; Bacillus subtilis subsp. subtilis',\n",
      " 'LineageEx': [{'TaxId': '131567', 'ScientificName': 'cellular organisms', 'Rank': 'no rank'}, {'TaxId': '2', 'ScientificName': 'Bacteria', 'Rank': 'superkingdom'}, {'TaxId': '1783272', 'ScientificName': 'Bacillati', 'Rank': 'kingdom'}, {'TaxId': '1239', 'ScientificName': 'Bacillota', 'Rank': 'phylum'}, {'TaxId': '91061', 'ScientificName': 'Bacilli', 'Rank': 'class'}, {'TaxId': '1385', 'ScientificName': 'Bacillales', 'Rank': 'order'}, {'TaxId': '186817', 'ScientificName': 'Bacillaceae', 'Rank': 'family'}, {'TaxId': '1386', 'ScientificName': 'Bacillus', 'Rank': 'genus'}, {'TaxId': '653685', 'ScientificName': 'Bacillus subtilis group', 'Rank': 'species group'}, {'TaxId': '1423', 'ScientificName': 'Bacillus subtilis', 'Rank': 'species'}, {'TaxId': '135461', 'ScientificName': 'Bacillus subtilis subsp. subtilis', 'Rank': 'subspecies'}],\n",
      " 'MitoGeneticCode': {'MGCId': '0', 'MGCName': 'Unspecified'},\n",
      " 'OtherNames': {'Misnomer': [], 'Teleomorph': [], 'Synonym': ['Bacillus subtilis 168', 'Bacillus subtilis subsp. subtilis 168', 'Bacillus subtilis subsp. subtilis str. BGSC 1A700'], 'Acronym': [], 'Anamorph': [], 'EquivalentName': [], 'GenbankAnamorph': [], 'CommonName': [], 'Name': [], 'GenbankSynonym': [], 'Misspelling': [], 'Includes': [], 'Inpart': []},\n",
      " 'ParentTaxId': '135461',\n",
      " 'PubDate': '2003/03/15 18:01:00',\n",
      " 'Rank': 'strain',\n",
      " 'ScientificName': 'Bacillus subtilis subsp. subtilis str. 168',\n",
      " 'TaxId': '224308',\n",
      " 'UpdateDate': '2015/07/30 14:00:50'}\n"
     ]
    }
   ],
   "source": [
    "from pprint import pprint\n",
    "# visualize first entry (check what's available)\n",
    "tax_id = list(cache)[0]\n",
    "pprint(dict(cache[tax_id][0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## compute some stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42924\n",
      "{'CreateDate': '2006/01/23 11:41:48',\n",
      " 'Division': 'Bacteria',\n",
      " 'GeneticCode': {'GCId': '11', 'GCName': 'Bacterial, Archaeal and Plant Plastid'},\n",
      " 'Lineage': 'cellular organisms; Bacteria; Bacillati; Bacillota; Bacilli; Bacillales; Staphylococcaceae; Staphylococcus; Staphylococcus aureus',\n",
      " 'LineageEx': [{'TaxId': '131567', 'ScientificName': 'cellular organisms', 'Rank': 'no rank'}, {'TaxId': '2', 'ScientificName': 'Bacteria', 'Rank': 'superkingdom'}, {'TaxId': '1783272', 'ScientificName': 'Bacillati', 'Rank': 'kingdom'}, {'TaxId': '1239', 'ScientificName': 'Bacillota', 'Rank': 'phylum'}, {'TaxId': '91061', 'ScientificName': 'Bacilli', 'Rank': 'class'}, {'TaxId': '1385', 'ScientificName': 'Bacillales', 'Rank': 'order'}, {'TaxId': '90964', 'ScientificName': 'Staphylococcaceae', 'Rank': 'family'}, {'TaxId': '1279', 'ScientificName': 'Staphylococcus', 'Rank': 'genus'}, {'TaxId': '1280', 'ScientificName': 'Staphylococcus aureus', 'Rank': 'species'}],\n",
      " 'MitoGeneticCode': {'MGCId': '0', 'MGCName': 'Unspecified'},\n",
      " 'OtherNames': {'Misnomer': [], 'Teleomorph': [], 'Synonym': [], 'Acronym': [], 'Anamorph': [], 'EquivalentName': ['Staphylococcus aureus subsp. aureus str. USA300', 'Staphylococcus aureus subsp. aureus strain USA300'], 'GenbankAnamorph': [], 'CommonName': [], 'Name': [], 'GenbankSynonym': [], 'Misspelling': [], 'Includes': [], 'Inpart': []},\n",
      " 'ParentTaxId': '1280',\n",
      " 'PubDate': '2006/02/11 18:00:11',\n",
      " 'Rank': 'strain',\n",
      " 'ScientificName': 'Staphylococcus aureus subsp. aureus USA300',\n",
      " 'TaxId': '367830',\n",
      " 'UpdateDate': '2018/11/23 17:31:12'}\n"
     ]
    }
   ],
   "source": [
    "# before stats: cache -> dict + convertion (easier to handle)\n",
    "data = {tax_id: dict(cache[tax_id][0]) for tax_id in unique_tax_ids + extra_tax_ids if cache[tax_id]}\n",
    "print(len(data))  # only valid entries\n",
    "# sample\n",
    "pprint(data[367830])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 42924/42924 [00:00<00:00, 453050.31it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'biotype',\n",
       " 'clade',\n",
       " 'class',\n",
       " 'family',\n",
       " 'genus',\n",
       " 'kingdom',\n",
       " 'no rank',\n",
       " 'order',\n",
       " 'pathogroup',\n",
       " 'phylum',\n",
       " 'serogroup',\n",
       " 'serotype',\n",
       " 'species',\n",
       " 'species group',\n",
       " 'species subgroup',\n",
       " 'strain',\n",
       " 'subclass',\n",
       " 'subfamily',\n",
       " 'subgenus',\n",
       " 'suborder',\n",
       " 'subspecies',\n",
       " 'superkingdom',\n",
       " 'tribe'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# available ranks\n",
    "ranks = set()\n",
    "for values in tqdm(data.values()):\n",
    "    if values:\n",
    "        for le in values.get(\"LineageEx\", {}):\n",
    "            if le:\n",
    "                ranks.add(le[\"Rank\"])\n",
    "ranks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TaxId': '561275',\n",
       " 'ScientificName': 'Mycobacterium tuberculosis variant bovis BCG str. Tokyo 172',\n",
       " 'OtherNames': {'Misnomer': [], 'Teleomorph': [], 'Synonym': ['Mycobacterium bovis BCG str. Tokyo 172'], 'Acronym': [], 'Anamorph': [], 'EquivalentName': ['Mycobacterium bovis BCG Tokyo 172', 'Mycobacterium bovis BCG strain Tokyo 172'], 'GenbankAnamorph': [], 'CommonName': [], 'Name': [], 'GenbankSynonym': [], 'Misspelling': [], 'Includes': [], 'Inpart': []},\n",
       " 'ParentTaxId': '33892',\n",
       " 'Rank': 'strain',\n",
       " 'Division': 'Bacteria',\n",
       " 'GeneticCode': {'GCId': '11', 'GCName': 'Bacterial, Archaeal and Plant Plastid'},\n",
       " 'MitoGeneticCode': {'MGCId': '0', 'MGCName': 'Unspecified'},\n",
       " 'Lineage': 'cellular organisms; Bacteria; Bacillati; Actinomycetota; Actinomycetes; Mycobacteriales; Mycobacteriaceae; Mycobacterium; Mycobacterium tuberculosis complex; Mycobacterium tuberculosis; Mycobacterium tuberculosis variant bovis; Mycobacterium tuberculosis variant bovis BCG',\n",
       " 'LineageEx': [{'TaxId': '131567', 'ScientificName': 'cellular organisms', 'Rank': 'no rank'}, {'TaxId': '2', 'ScientificName': 'Bacteria', 'Rank': 'superkingdom'}, {'TaxId': '1783272', 'ScientificName': 'Bacillati', 'Rank': 'kingdom'}, {'TaxId': '201174', 'ScientificName': 'Actinomycetota', 'Rank': 'phylum'}, {'TaxId': '1760', 'ScientificName': 'Actinomycetes', 'Rank': 'class'}, {'TaxId': '85007', 'ScientificName': 'Mycobacteriales', 'Rank': 'order'}, {'TaxId': '1762', 'ScientificName': 'Mycobacteriaceae', 'Rank': 'family'}, {'TaxId': '1763', 'ScientificName': 'Mycobacterium', 'Rank': 'genus'}, {'TaxId': '77643', 'ScientificName': 'Mycobacterium tuberculosis complex', 'Rank': 'species group'}, {'TaxId': '1773', 'ScientificName': 'Mycobacterium tuberculosis', 'Rank': 'species'}, {'TaxId': '1765', 'ScientificName': 'Mycobacterium tuberculosis variant bovis', 'Rank': 'biotype'}, {'TaxId': '33892', 'ScientificName': 'Mycobacterium tuberculosis variant bovis BCG', 'Rank': 'no rank'}],\n",
       " 'CreateDate': '2008/09/17 12:37:44',\n",
       " 'UpdateDate': '2018/11/23 17:53:12',\n",
       " 'PubDate': '2008/10/15 18:00:10'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# interesting sample (contains special ranks)\n",
    "data[561275]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get all valid tax_id in dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def clean_tax_id(tax_id):   \n",
    "    try:        \n",
    "        return int(tax_id)\n",
    "    except ValueError:\n",
    "        return None\n",
    "    \n",
    "tax_ids = df[\"tax_id\"].apply(clean_tax_id).dropna().astype(int).to_list()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1943014/1943014 [00:01<00:00, 1001868.29it/s]\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "linstats = defaultdict(int)\n",
    "for tax_id in tqdm(tax_ids):\n",
    "    if tax_id not in data:\n",
    "        continue\n",
    "    tdata = data[tax_id]\n",
    "    if tdata:\n",
    "        for lin in tdata[\"LineageEx\"]:\n",
    "            linstats[lin[\"TaxId\"]] += 1       \n",
    "dfcount = pd.DataFrame.from_dict(linstats, orient=\"index\", columns=[\"count\"]).sort_values(\"count\", ascending=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>name</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>131567</th>\n",
       "      <td>1941392</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>no rank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1941392</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>superkingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3379134</th>\n",
       "      <td>1316225</td>\n",
       "      <td>Pseudomonadati</td>\n",
       "      <td>kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1224</th>\n",
       "      <td>1184764</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>phylum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236</th>\n",
       "      <td>1073702</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1567502</th>\n",
       "      <td>1</td>\n",
       "      <td>Brucella suis bv. 5</td>\n",
       "      <td>no rank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2649111</th>\n",
       "      <td>1</td>\n",
       "      <td>unclassified Atopobacter</td>\n",
       "      <td>no rank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34019</th>\n",
       "      <td>1</td>\n",
       "      <td>Liberibacter</td>\n",
       "      <td>genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344063</th>\n",
       "      <td>1</td>\n",
       "      <td>Oryzihumus</td>\n",
       "      <td>genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120376</th>\n",
       "      <td>1</td>\n",
       "      <td>Salana</td>\n",
       "      <td>genus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6566 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           count                      name          rank\n",
       "131567   1941392        cellular organisms       no rank\n",
       "2        1941392                  Bacteria  superkingdom\n",
       "3379134  1316225            Pseudomonadati       kingdom\n",
       "1224     1184764            Pseudomonadota        phylum\n",
       "1236     1073702       Gammaproteobacteria         class\n",
       "...          ...                       ...           ...\n",
       "1567502        1       Brucella suis bv. 5       no rank\n",
       "2649111        1  unclassified Atopobacter       no rank\n",
       "34019          1              Liberibacter         genus\n",
       "344063         1                Oryzihumus         genus\n",
       "120376         1                    Salana         genus\n",
       "\n",
       "[6566 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# export file : dfcount.xlsx\n",
    "\n",
    "def get_attr(tax_id, attr):\n",
    "    tax_id = int(tax_id)\n",
    "    if tax_id in data:\n",
    "        return data[tax_id].get(attr, \"\")\n",
    "    return \"\"\n",
    "dfcount[\"name\"] = dfcount.index.map(lambda tax_id: get_attr(tax_id, \"ScientificName\"))\n",
    "dfcount[\"rank\"] = dfcount.index.map(lambda tax_id: get_attr(tax_id, \"Rank\"))\n",
    "dfcount.to_excel(\"out/dfcount.xlsx\")\n",
    "dfcount"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## count by rank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import networkx as nx\n",
    "# gtaxo = nx.DiGraph()\n",
    "# for tax_id, values in data.items():\n",
    "#     gtaxo.add_node(tax_id)\n",
    "\n",
    "# gtaxo"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "microtaxo-SPpTyrCZ-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
